### PR TITLE
Updated shortest job first and Process

### PR DESCRIPTION
**Pull request by FMock**
SJF algorithm requires sorting the list of Processes by the estimated arrival time. Therefore I updated the Process class to implement the Comparator interface and added the required compare method. An empty constructor and a public getEstRunTime  method were also required and added.
In the ShortestJobFirst class I completed the first step of the algorithm which is to sort the Processes by estimated run time with Collections.sort(processes, new Process())

### CODE REVIEW

File: Process.java [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
 	private String name;
 	private int priority;
 	private float arrivalNum;
-	private float runTime;
+	private float runTime; //expected run time
 	private float timeLeft;
 	private float startTime;
 	private float endTime;
 	
+	public Process(){}
```
Comment (flip40) [inline Process.java]: Do we ever need to create empty processes? Not a big deal if we have the option though I guess.

File: Process.java [hunk 2]
```diff
@@ -75,6 +81,25 @@ public void reset() {
 	//	output.outputResults((endTime - startTime), (endTime - startTime - runTime), 0);
 	//}
 	
+
```
Comment (flip40) [inline Process.java]: I just made a PR at #12 that does a similar function. I feel it is a bit better implementation, as it allows sorting by Arrival Time, Priority, Total Run Time, or Time Remaining. By having Process itself implement comparable, you only have one comparison option. Implementing comparator adds a second, but we may need more.

File: Process.java (hunk 1 above)
Comment (FMock) [inline Process.java]: The empty Process constructor is so SJF can sort the list of Processes by arrival time - Collections.sort(processes, new Process()) 
-Frank M 

----- Original Message -----

Comment (flip40) [inline Process.java]: I disagree with this. Instantiating an empty Process object as a comparator object for the other Processes is a bit roundabout. It is like instantiating an Integer to sort other Integers. Process as a comparable makes sense, it is self comparing (though for our purposes, falls short due to a single comparison method), but Process should _not_ be a comparator. I would still argue for #12 which uses private comparator classes that are accessible through their respective static function of the Process class.

**Comment by flip40** [comment]
I think we agreed to use #12 for process sorting (from email). Again, apologies for not realizing you too were attempting to do the same thing. I should have made a note that I was working on something.

Assuming we use #12 I think this needs the process changes removed from this PR, as well as the SJF class needs to be slightly updated to fit with the new comparison method.

By the way, if you want to add the changes I found this, which seems to properly follow how to merge master changes into your own branch (prefering the merge method over the rebase): http://stackoverflow.com/a/3877000

**Comment by FMock** [comment]
I tried to move the changes to my branch but got an error with the merge that I'm not sure how to correct.
$ git merge origin/master
error: merge is not possible because you have unmerged files.
hint: Fix them up in the work tree, and then use 'git add/rm <file>'
hint: as appropriate to mark resolution and make a commit.
fatal: Exiting because of an unresolved conflict.

I guess I'll just copy the code from master or edit (where ever the new code is)

**Comment by flip40** [comment]
You are getting an error because you have changed the Process.java class and it conflicts with the changes I made and then merged into master. You needed to revert your Process.java file to the master branch version. For simplicity sake though, you could just start a new branch and copy your changes there and merge that if that is easier. At the moment, you might have to figure out how to stop the merge and clean the branch of the merge files. If you know how to do that, this link indicates how you would go about getting a specific file version into your branch: http://stackoverflow.com/a/215731

**Comment by flip40** [comment]
Seems like you fixed it, +1

**Comment by FMock** [comment]
I believe I have implemented SJF correctly. The output looks good. Take a look at my code and run it when you guys get a chance. I had to add a getEndTime() method to Process.java
I think ShortestJobFirst.java and Process.java can merge back to the master branch.
Here's the output
[SJF_OUTPUT.txt](https://github.com/CS149-3/homework2/files/139422/SJF_OUTPUT.txt)

**Comment by flip40** [comment]
We need the process to return a Result object, it must be filled out to meet the output requirements defined on the homework.

Also you are not taking into account the real time requirements (i.e. arrival time and inability for the scheduler to sort processes except at the start/end of a quantum).

For example your output shows that Process F, who's arrival time is 99.749245 started BEFORE it arrived, at 0.0. After this, Process P was ran, who arrival time was 54.579117 (earlier than Process F) and yet was ran after Process F, starting at 0.3.

**Comment by FMock** [comment]
Shortest Job First can be merged. The algorithm isn't implemented perfectly but at this point I just need to hand it over to the main branch. I had to  add some setters and getters to Result and Process so there are conflicts with the main branch. John can you please merge to main?
***please remove the `package homework2;` statement at the top of the files ***
Here are my 5 Result object outputs in file form:
[sjf_output_001.txt](https://github.com/CS149-3/homework2/files/141828/sjf_output_001.txt)
[sjf_output_002.txt](https://github.com/CS149-3/homework2/files/141825/sjf_output_002.txt)
[sjf_output_003.txt](https://github.com/CS149-3/homework2/files/141824/sjf_output_003.txt)
[sjf_output_004.txt](https://github.com/CS149-3/homework2/files/141827/sjf_output_004.txt)
[sjf_output_005.txt](https://github.com/CS149-3/homework2/files/141826/sjf_output_005.txt)

**Pull request merged**